File: src/main/java/com/wesker926/leetcode/algorithms/p0563/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/11/18
// @description: dfs
package s1

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findTilt(root *TreeNode) int {
	gradient, _ := dfs(root)
	return gradient
}

func dfs(root *TreeNode) (int, int) {
	if root == nil {
		return 0, 0
	}

	lg, ls := dfs(root.Left)
	rg, rs := dfs(root.Right)
	return lg + rg + abs(ls-rs), ls + rs + root.Val
}

func abs(i int) int {
	if i >= 0 {
		return i
	}
	return -i
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0563/s1/Solution.java
package com.wesker926.leetcode.algorithms.p0563.s1;

/**
 * @author wesker.gh
 * @date 2021/11/18
 * @description TODO @wesker.gh
 */
public class Solution {

    private int gradient = 0;

    public int findTilt(TreeNode root) {
        dfs(root);
        return gradient;
    }

    private int dfs(TreeNode node) {
        if (node == null) {
            return 0;
        }

        int ls = dfs(node.left);
        int rs = dfs(node.right);

        gradient += Math.abs(ls - rs);
        return ls + rs + node.val;
    }

    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}
